server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java

```diff
@@ -0,0 +1,343 @@
+                SearchShardTarget target = result.getSearchShardTarget();
+                processedShards.add(new SearchShard(target.getClusterAlias(), target.getShardId()));
+            }
+            progressListener.onPartialReduce(processedShards, topDocsStats.getTotalHits(), reducedAggs, numReducePhases);
+        }
+    }
+
+    private List<InternalAggregations> consumeAggs(List<QuerySearchResult> results) {
+        if (hasAggs) {
+            List<InternalAggregations> aggsList = new ArrayList<>();
+            if (reducedAggs != null) {
+                aggsList.add(reducedAggs.expand());
+                reducedAggs = null;
+            }
+            for (QuerySearchResult result : results) {
+                aggsList.add(result.consumeAggs().expand());
+            }
+            return aggsList;
+        } else {
+            return null;
+        }
+    }
+
+    private List<TopDocs> consumeTopDocs(List<QuerySearchResult> results) {
+        if (hasTopDocs) {
+            List<TopDocs> topDocsList = new ArrayList<>();
+            if (reducedTopDocs != null)  {
+                topDocsList.add(reducedTopDocs);
+            }
+            for (QuerySearchResult result : results) {
+                TopDocsAndMaxScore topDocs = result.consumeTopDocs();
+                topDocsStats.add(topDocs, result.searchTimedOut(), result.terminatedEarly());
+                setShardIndex(topDocs.topDocs, result.getShardIndex());
+                topDocsList.add(topDocs.topDocs);
+            }
+            return topDocsList;
+        } else {
+            return null;
+        }
+    }
+
+    private synchronized void onFatalFailure(Exception exc) {
+        onPartialMergeFailure.accept(exc);
+        fatalFailure.compareAndSet(null, exc);
+        MergeTask task = runningTask.get();
+        if (task != null) {
+            task.cancel();
+        }
+        runningTask.compareAndSet(task, null);
+        queue.stream().forEach(MergeTask::cancel);
+        queue.clear();
+        reducedTopDocs = null;
+        reducedAggs = null;
+    }
+
+    private boolean hasFailure() {
+        return fatalFailure.get() != null;
+    }
+
+    synchronized boolean hasPendingMerges() {
+        return queue.isEmpty() == false || runningTask.get() != null;
+    }
+
+    synchronized void tryExecute(MergeTask task) {
+        queue.add(task);
+        tryExecuteNext();
+    }
+
+    private void tryExecuteNext() {
+        final MergeTask task;
+        synchronized (this) {
+            if (queue.isEmpty()
+                    || runningTask.get() != null
+                    || hasFailure()) {
+                return;
+            }
+            runningTask.compareAndSet(null, queue.peek());
+            task = queue.poll();
+        }
+        executor.execute(new AbstractRunnable() {
+            @Override
+            protected void doRun() throws Exception {
+                List<QuerySearchResult> toConsume = task.consumeBuffer();
```

Comment:
Do you see the task being consumed twice? It is probably worth a comment explaining why.

Reply:
If the task is cancelled, we nullify the buffer in the task. That's why we need this check, I added a comment in https://github.com/elastic/elasticsearch/pull/58461/commits/1fa48801f433588a8a0de2e64851c1e145cf0640